Refactor Number.prototype methods toFixed, toExponential, toPrecision

szilagyiadam opened the pull request:
JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -651,89 +613,106 @@ static ecma_value_t
+  int32_t exponent;
 
-  lit_utf8_byte_t *p = buff;
+  num_of_digits = ecma_number_to_binary_floating_point_number (this_num, digits, &exponent);
```
I think this function should be removed as well. The other functions all use `ecma_number_to_decimal` to convert the number to digits, we should use that here as well.

The names of some of the variables should also be revised, they seem to store different values than what they imply. For example currently calling `ecma_number_to_binary_floating_point_number ` on the number `5` will result in `num_digits = 0, digits = "5", exponenet = 1`, which makes no sense at all.

After switching to `ecma_number_to_decimal` we should double check the variables and name them appropriately. Maybe even `ecma_number_to_decimal` should be renamed/updated a bit so that it's purpose and the meaning of the returned values are clearer.

szilagyiadam commented:
Updated the PR!
Also refactored toExponential and toPrecision, now every method uses one common method.
Note: I don't have a solution for the toFixed method where i can use ecma_number_to_decimal yet.

szilagyiadam commented:
> Updated the PR!
> Also refactored toExponential and toPrecision, now every method uses one common method.
> Note: I don't have a solution for the toFixed method where i can use ecma_number_to_decimal yet.

Did an update so i use ecma_number_to_decimal in all of the three methods

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -574,266 +472,301 @@ typedef enum
+    if (mode == NUMBER_ROUTINE_TO_FIXED)
     {
-      digits[i] = '0';
+      arg_int = 0;
```
If the argument was undefined then this should already be 0.

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -574,266 +472,301 @@ typedef enum
+    return to_integer;
   }
 
+  /* Argument boundary checks */
```
These ranges have changed in the recent standard, please update them.

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -574,266 +472,301 @@ typedef enum
+  /* toExponent routine and toPrecision cases where the exponent > precision or exponent < -5 */
+  if (mode == NUMBER_ROUTINE_TO_EXPONENTIAL
+      || (mode == NUMBER_ROUTINE_TO_PRECISION
+          && (exponent < -5 || exponent > arg_int)))
```
This range seems to be different in the standard. Please add tests that cover the edge cases and make sure they are correct.

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -849,120 +782,9 @@ static ecma_value_t
-  JMEM_FINALIZE_LOCAL_ARRAY (buff);
-
-  return ret_value;
+  return ecma_builtin_number_prototype_object_to_number_convert (this_num,
```
We could call these directly from the dispatcher.

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -574,266 +472,301 @@ typedef enum
-  }
-  else
+  /* Handle undefined argument */
+  if (ecma_is_value_undefined (arg))
```
The required actions when the argument is undefined also changed slightly n the standard, please double check them.

dbatyai commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -574,396 +472,262 @@ typedef enum
+      && exponent > 21)
   {
-    frac_digits = (int32_t) num_digits - 1;
+    ecma_value_t result_num = ecma_builtin_number_prototype_object_to_string (this_num, NULL, 0);
```
How about destroying the builder first, and returning with the result of the to_string call?

rerobika commented on jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c:
```diff
@@ -1007,15 +771,21 @@ ecma_builtin_number_prototype_dispatch_routine (uint16_t builtin_routine_id, /**
+                                                                     routine_arg_1,
+                                                                     NUMBER_ROUTINE_TO_EXPONENTIAL);
     }
     case ECMA_NUMBER_PROTOTYPE_TO_PRECISION:
```
Please merge these 3 cases. the number routine mode can be calculated like this:

NUMBER_ROUTINE_TO_FIXED + (builtin_id - ECMA_NUMBER_PROTOTYPE_TO_FIXED);
